### PR TITLE
feat: Add `is_noop` predicate to `MCMCPropagator` to skip patch and density evaluation for empty moves

### PR DESCRIPTION
**Pull request by n-gao**
Adds an optional `is_noop` predicate to `MCMCPropagator` that short-circuits patch construction and log-probability evaluation when a proposed move is guaranteed to leave the state unchanged. When the predicate returns true, the propagator skips calling `patch_fn` and `log_probability_ratio_fn`, while still consuming the same random keys and running scheduler updates to keep trajectories reproducible. The branch is implemented with `jax.lax.cond`, so random key consumption and scheduler state remain identical whether or not the predicate fires, ensuring that enabling `is_noop` does not alter the trajectory of non-empty moves.

The `Probe` protocol is also relaxed so that its payload type parameter `P` is no longer bounded to `Patch[Any]`. This allows `Probe` to describe functions that inspect raw proposal changes before a patch is constructed, which is the typical use case for `is_noop` predicates.

Three tests cover the new behaviour: one verifies that `patch_fn` and `log_probability_ratio_fn` are not called for empty moves while acceptance and scheduling still advance correctly; one confirms that a mixed-move propagator with `is_noop` produces an identical trajectory to the baseline without it; and one checks that assertions in the proposal and active evaluation branch are still enforced as expected.

Runtime assertion handling reuses each interpreted conditional branch for both shape discovery and execution, including inside disabled-JIT contexts. A nested-conditional test checks one interpretation per assertion, failure propagation, and gradients.

### Performance measurements

Synthetic density difference over 4,194,304 elements (loaded) or 262,144 elements (small API), with empty and active proposals; acceptance and scheduler updates remain enabled.

| Workload | Cold trace, parent → PR (ms) | Runtime, parent → PR (ms) | Speedup |
|---|---:|---:|---:|
| CPU, 4.2M, empty | 22.5 → 28.6 | 69.2 → 6.84 | 10.11× |
| CPU, 4.2M, active | 22.3 → 26 | 69.4 → 69.7 | 1.00× |
| L4, 4.2M, empty | 22.7 → 26 | 1.38 → 0.596 | 2.32× |
| L4, 4.2M, active | 23 → 26 | 1.39 → 1.54 | 0.90× |
| CPU, 262k, empty | 24.2 → 27.2 | 3.6 → 0.624 | 5.77× |
| CPU, 262k, active | 22.6 → 26.1 | 3.58 → 3.97 | 0.90× |
| L4, 262k, empty | 22.6 → 27.1 | 0.487 → 0.596 | 0.82× |
| L4, 262k, active | 22.6 → 26.9 | 0.508 → 0.56 | 0.91× |

Loaded L4 activity: empty 75.5 → 51.5%; active 75 → 77.5%.
Small API timings include dispatch and tree reconstruction: three cold traces and 25 warmed synchronous calls per process. Loaded numerical timings use flat array inputs/outputs, at least 3 s and 10 warmed synchronous calls per phase. GPU activity comes from 200 ms `nvidia-smi` samples after the first second; it is not arithmetic occupancy.

Production assertion tracing: four boxes of 50 charged CO₂ molecules, LJ+tail and Ewald. Empty-move detection is wired into this application later in #313, so this comparison isolates the conditional-assertion tracing change.

| Workload | Cold trace, parent → PR (s) | Runtime, parent → PR (ms) | Speedup |
|---|---:|---:|---:|
| CPU charged proposal | 2.91 → 2.71 | 4.02 → 4.63 | 0.87× |

This production row uses two cold traces and ten warmed blocks of 16 proposals per process. Cached energies match full recomputation.

Parent `f493ad3` → PR `1d5df45`. JAX 0.10.0 / Python 3.14, float64; Xeon 2.20 GHz (CPU affinity 0–7), NVIDIA L4. Two fresh processes per revision in reversed order; medians of process medians. Trace excludes lowering and XLA compilation. Speedup is parent/PR; below 1× is slower.

Empty proposals benefit when density evaluation is large. Active guards still add overhead, and the small L4 cases regress; production exchange detection is measured separately in #313.

### CODE REVIEW

**Comment by n-gao** [comment]
* **#313** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/313?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 2  dependent PRs ([#314](https://github.com/cusp-ai-oss/kUPS/pull/314) <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/314?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#315](https://github.com/cusp-ai-oss/kUPS/pull/315) <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/315?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#312** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/312?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#311** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/311?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/311?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#310** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/310?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Review by felixmusil — APPROVED** [review]

**Comment by n-gao** [comment]
### Merge activity

* **Sep 16, 9:38 AM UTC**: @n-gao merged this pull request with [Graphite](https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/311).
